### PR TITLE
[docs] explain input equality.

### PR DESCRIPTION
**Pull request by Laimiux**
## What
Input equality is a common issue that comes up in PR reviews.

### CODE REVIEW

**Comment by carrotkite** [comment]
### JaCoCO Code Coverage 76.85% :white_check_mark:
| Class | Covered | Meta | Status |
|:---|:---:|:---:|:---:|


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: docs/input.md [hunk 1]
```diff
@@ -47,3 +47,131 @@ itemDetailFormula
...
+  val eventObservable: Observable<Event>
+)
+```
+
+**Don't:** create a new observable within `Formula.evaluate`
+```kotlin
+val input = MyInput(
+    eventObservable = relay.map { Event() }
+)
+```
```
Comment (Jawnnypoo) [inline docs/input.md:134]: I don't think this is the correct example - this isn't creating a new observable within `evaluate`

Comment (Jawnnypoo) [reply docs/input.md:134]: Should be:
```
override fun evaluate(...) {
    val input = MyInput(
        eventObservable = eventObservable
    )
}
```

File: docs/input.md [hunk 2]
```diff
@@ -47,3 +47,131 @@ itemDetailFormula
...
+override fun initialState(input: Input) = State(
+  eventObservable = input.eventObservable.map { Event() }
+)
+
+override fun evaluate(...) {
+    val input = MyInput(
+        eventObservable = state.eventObservable
+    )
+}
+```
```
Comment (Jawnnypoo) [inline docs/input.md:163]: Do we still want to use `State` if we are directly taking the Observable from `Input`, not modifying it and using its raw result within `events`? ie. 
```
   data class Input(
        val complementaryMatchesEventRelay: Observable<Boolean>
   )

    override fun evaluate(
        input: Input,
        state: State,
        context: FormulaContext<State>
    ): Evaluation<ICNotificationTrayRenderModel> {
        return Evaluation(
            renderModel = renderModel,
            updates = context.updates {
                events(input.complementaryMatchesEventRelay) {
                    transition(state.copy(complementaryMatches = it))
                }
            }
        )
    }
```

Comment (Laimiux) [reply docs/input.md:163]: Using `Observable` directly from the `Input` is fine, the parent has to ensure equality.

**Comment by Jawnnypoo** [comment]
A scenario that I run into quite frequently is that in order to create a `Stream` via `RxStream.fromObservable`, I need to use the `Input`. In this case, is it best to create an initial Stream with `Observable.empty()`, and then update it to the correct `Stream` in `onInputChange`, or is there some other way to accomplish this?

**Comment by Laimiux** [comment]
> A scenario that I run into quite frequently is that in order to create a `Stream` via `RxStream.fromObservable`, I need to use the `Input`. In this case, is it best to create an initial Stream with `Observable.empty()`, and then update it to the correct `Stream` in `onInputChange`, or is there some other way to accomplish this?

Could you show an example of how you use the `Input`?

If you rely on data coming from the `Input`, you can use `cacheKey` to configure the right invalidation policy, for example 
```kotlin
RxStream.fromObservable(input.itemId) { repo.fetchitem(input.itemId) }
```

**Pull request merged**